text/0001-script-based-tokenizer.md

```diff
@@ -0,0 +1,187 @@
+
+### Explanation
+
+The tokenizer serves as a proxy for other tokenizers, specialized in the language detected by the Tokenizer. It is instantiated with a String, and is then polled for tokens until it's delepted:
+```rust
+let tokenizer = Tokenizer::new("The quick brown fox jumps over the lazy dog");
+assert_eq!(tokenizer.next().txt(), "The")
+assert_eq!(tokenizer.next().txt(), " ")
+assert_eq!(tokenizer.next().txt(), "quick")
+```
+
+### Impact on documentation
+
+This feature should not impact the meilisearch user documentation,
+in future versions we will probably provide a way to configure tokenizer and this will be discussed in a new specification.
+
+## Technical Specifications
+
+### Architecture
+
+As the actual tokenizer, this tokenizer will be a standolone library,
+and will be used by meilisearch like another one.
+We decided to put it in an other repository to be able to use it in meilisearch/Meilisearch and in the future core-engine.
+
+### Implementation Details
+
+We could have 2 approach of tokenization based on the script,
+the main difficulty is to handle multi-script documents
+which need to segment each part of script and tokenize them differently,
+theses 2 approach are:
+- first iterate over the whole document, splitting it by script, and after, tokenize each part with specialized lexer
+- detect the script of the start of the document, tokenize with specialized lexer while script is the same, switch lexer if the script change
+
+#### Tokenizer
+
+```rust
+use crate::token::Token;
+use crate::internal_tokenizer::InternalTokenizer;
+
+struct Tokenizer<'a> {
+    /// script specialized tokenizer, this can be switched during
+    /// document tokenization if the document contains several scripts
+    current_tokenizer: Option<Box<dyn InternalTokenizer<'a>>>,
+    /// current character index in the document
+    current_char_index: u64,
+    /// reference on the document content
+    inner: &'a str,
+}
+
+impl<'a> Tokenizer<'a> {
+    /// create a new tokenizer detecting script
+    /// and chose the specialized internal tokenizer
+    fn new(inner: &'a str) -> Self { unimplemented!() }
+}
+
+impl<'a> Iterator for Tokenizer<'a> {
+    type Item = Token<'a>;
+
+    /// return the next token calling the `next` method of the internal tokenizer,
+    /// if the internal tokeizer return None, the function try to redetect script and chose a new tokenizer,
+    /// if no iternal tokenizer is chosen, the method return None
+    fn next(&mut self) -> Option<Self::Item> { unimplemented!() }
+}
+```
+
+#### Token
+
+```rust
+/// script of a token (https://docs.rs/whatlang/0.10.0/whatlang/enum.Script.html)
+pub type Script = whatlang::Script;
+
+/// atomic item returned by `Tokenizer::next()`
+pub struct Token<'a> {
+    word: Lexeme<'a>,
+    script: Script,
+}
```

Comment:
@Kerollmops @LegendreM 
We want to allow for overlapping tokens (i.e tokens that may refer to the same word position). There are multiple solutions to that:
- (prefered) add a `word_position` field to the WordSlice, multiple `WordSlice` can have the same word index.
- Make `Lexme` contain a Cow<[WordSlice]> that contains all the word slices for the same word position, this implies that token's `txt` method returns a collection of `&str` which may not be very ergonomic?
- Make `Token` contain a `Cow<[Lexeme]>`, similar to before, a bit more general, same issue